src/main/scala/scala/collection/decorators/IteratorDecorator.scala

```diff
@@ -72,4 +72,47 @@ class IteratorDecorator[A](val `this`: Iterator[A]) extends AnyVal {
     loop(immutable.List.empty)
   }
 
+  /**
+    * Constructs an iterator where consecutive elements are accumulated as
+    * long as the output of f for each element doesn't change.
+    * <pre>
+    * Vector(1,2,2,3,3,3,2,2)
+    *     .iterator
+    *     .groupUntilChanged(identity)
+    *     .toList
+    * </pre>
+    * produces
+    * <pre>
+    * List(Seq(1),
+    *      Seq(2,2),
+    *      Seq(3,3,3),
+    *      Seq(2,2))
+    * </pre>
+    * @param f the function to compute a key for an element
+    * @tparam K the type of the computed key
+    * @return an iterator of sequences of the consecutive elements with the
+    *         same key in the original iterator
+    */
+  def groupUntilChanged[K](f: A => K): Iterator[Iterable[A]] =
+    new AbstractIterator[Seq[A]] {
+      private var hd: Option[A] = readHead()
```

Comment:
Is it necessary to have such a variable ? It's used in `next()` and `hasNext`. For `next()`, the match could be directly on `readHead()` whereas `hasNext` could rely on `this.hasNext` instead. This would delay the `next` action to when `next` is actually called, at least for the first call on `next`

Reply:
I don't see a way around the variable. We need to peek ahead in the underlying iterator in order to know if we should continue accumulating or if we should start another group. Having read one ahead, it has to be stored locally or it would be lost.

Reply:
True, I did not see that point. The `seq.toVector` can only be done if we know that the next element changes key. Too bad we could not just rely on the state of the underlying Iterable to do that.